sites/kit.svelte.dev/src/lib/search/SearchBox.svelte

```diff
@@ -0,0 +1,416 @@
+<script>
+	import { onMount } from 'svelte';
+	import { Icon } from '@sveltejs/site-kit';
+	import flexsearch from 'flexsearch';
+	import { afterNavigate } from '$app/navigation';
+	import { searching, query, recent } from './stores.js';
+	import { focusable_children, trap } from '../actions/focus.js';
+
+	let modal;
+
+	let results = [];
+	let backspace_pressed;
+
+	let index;
+	let lookup;
+
+	onMount(async () => {
+		const response = await fetch('/content.json');
+		const { blocks } = await response.json();
+
+		index = new flexsearch.Index({
+			tokenize: 'forward'
+		});
+
+		lookup = new Map();
+
+		let time = Date.now();
+		for (const block of blocks) {
+			const title = block.breadcrumbs[block.breadcrumbs.length - 1];
+			lookup.set(block.href, {
+				title,
+				href: block.href,
+				breadcrumbs: block.breadcrumbs.slice(0, -1),
+				content: block.content
+			});
+			index.add(block.href, `${title} ${block.content}`);
+
+			// poor man's way of preventing blocking
+			if (Date.now() - time > 25) {
+				await new Promise((f) => setTimeout(f, 0));
+				time = Date.now();
+			}
+		}
+
+		update();
+	});
+
+	afterNavigate(() => {
+		// TODO this also needs to apply when only the hash changes
+		// (should before/afterNavigate fire at that time? unclear)
+		$searching = false;
+	});
+
+	function update() {
+		results = (index ? index.search($query) : []).map((href) => lookup.get(href));
+	}
+
+	function escape(text) {
+		return text.replace(/</g, '&lt;').replace(/>/g, '&gt;');
+	}
+
+	function excerpt(content, query) {
+		const index = content.toLowerCase().indexOf(query.toLowerCase());
+		if (index === -1) {
+			return content.slice(0, 100);
+		}
+
+		const prefix = index > 20 ? `…${content.slice(index - 15, index)}` : content.slice(0, index);
+		const suffix = content.slice(
+			index + query.length,
+			index + query.length + (80 - (prefix.length + query.length))
+		);
+
+		return (
+			escape(prefix) +
+			`<mark>${escape(content.slice(index, index + query.length))}</mark>` +
```

Comment:
Marking up content like this is having the screen readers I tested (NVDA and VoiceOver) break up the words, which could be confusing. For example, if I search "rou", the first result's heading is output as `<mark>Rou</mark>ting`, which is read by NVDA as "highlighted    Rou  out of highlighted  ting". VoiceOver doesn't announce "highlight", but does separate into separate words, i.e. "Rou ting".

I wonder if it would be a good idea to output the marked-up strings twice: once with the `<mark>` elements and once without. The marked version would be hidden from screen readers with aria-hidden, and the unmarked version would be hidden from sighted users [with CSS](https://piccalil.li/quick-tip/visually-hidden/). This way, screen readers would have the full string read out without being broken up.

Reply:
ah, interesting. the only real reason to use `<mark>` is for semantic correctness, which I had naively assumed would be the more accessible choice. if it's actually harmful to accessibility to mark incomplete words, then maybe it should just be a `<span class="highlight">` instead?

Reply:
Upon further reflection, I think this is fine to leave as-is:

1. `<mark>` is the more semantically correct option here
1. NVDA lets users [configure this behavior](https://www.nvaccess.org/files/nvda/documentation/userGuide.html#DocumentFormattingSettings), so users can turn off the "highlight" announcement if they want. 
1. Since I'm not certain that hearing incomplete words like this is an accessibility issue, we probably shouldn't over-optimize for that assumption. 

> To a developer or user who is used to the way JAWS or NVDA announces content, the behavior of VO might seem fragmented and annoying, but to someone who regularly uses VO, this pattern of speech output is usual, and may be the expected or even preferred pattern. It’s not necessarily accurate to call this a usability problem at all, and it’s not safe to assume that the typical behavior of an assistive technology is something that needs to be “fixed”, simply because it’s inconsistent with the behavior of similar technologies. In any case, this quirk only exists at all when navigating with the default side-swipe gesture pattern, it doesn’t occur when navigating using rotor lines navigation, or when VO reads continuously (but the problems caused by overriding semantics do still occur in all circumstances).
> &mdash;[TPGI](https://www.tpgi.com/using-the-text-role/)